### PR TITLE
Implement custom deserialization and validation logic

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/dotnet/tye/issues/232.

### CODE REVIEW

File: src/Microsoft.Tye.Core/ApplicationFactory.cs [hunk 1]
```diff
@@ -244,84 +243,5 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
 
             return builder;
         }
-
-        private static void ValidateConfigApplication(ConfigApplication config)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ApplicationFactory.cs:248]: Why did you remove this?

File: src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs [hunk 2]
```diff
@@ -28,5 +33,139 @@ public class ConfigApplication
...
+                throw new TyeYamlException(
+                    "Configuration validation failed." + Environment.NewLine +
+                    string.Join(Environment.NewLine, results.Select(r => r.ErrorMessage)));
+            }
+
+            foreach (var extension in config.Extensions)
+            {
+                if (!extension.TryGetValue("name", out var name) || string.IsNullOrWhiteSpace(name as string))
+                {
+                    throw new TyeYamlException(CoreStrings.ExtensionMustProvideAName);
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs:55]: oh no resources. lol

Do you think there's a really important reason we use resources?

File: src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs [hunk 3]
```diff
@@ -28,5 +33,139 @@ public class ConfigApplication
...
+                            $"Volume '{volume.Source}' of service '{service.Name}' validation failed." + Environment.NewLine +
+                            string.Join(Environment.NewLine, results.Select(r => r.ErrorMessage)));
+                    }
+                }
+            }
+
+            foreach (var ingress in config.Ingress)
+            {
+                // We don't currently recurse into ingress rules or ingress bindings right now.
+                // There's nothing to validate there.
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs:120]: Comment seems stale now

File: src/Microsoft.Tye.Core/ConfigModel/ConfigIngress.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@ namespace Microsoft.Tye.ConfigModel
     public class ConfigIngress
     {
         [Required]
-        public string Name { get; set; } = default!;
+        public string? Name { get; set; } = default!;
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/ConfigModel/ConfigIngress.cs:13]: Remove `default!` if you're going to make these nullable. 

The point of how this worked in the past:

The nullability contract + `[Required]` reflect the **valid** state of the object. That's what `default!` is used. If you're going to treat the properties as nullable, there's no need to initialized them explicitly.

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj [hunk 5]
```diff
@@ -32,6 +32,12 @@
     <Content Include="Templates\**" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
 
+  <ItemGroup>
+    <EmbeddedResource Update="CoreStrings.resx" EmitFormatMethods="true">
+      <Generator></Generator>
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:37]: Isn't this supposed to be set to something?

File: src/Microsoft.Tye.Core/Serialization/TyeYamlException.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            : this(start,  message, null)
+        {
+        }
+
+        public TyeYamlException(Mark start, string message, Exception? innerException)
+            : base($"Error parsing tye.yaml: ({start.Line}, {start.Column}): {message}", innerException)
+        {
+        }
+
+        public TyeYamlException(string message, Exception inner)
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/TyeYamlException.cs:27]: `Exception?`

File: src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Tye.ConfigModel;
+using YamlDotNet.RepresentationModel;
+
+namespace Tye.Serialization
+{
+    public static class YamlConfigApplicationHelpers
+    {
+
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:12]: @jamesnk

File: src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Tye.ConfigModel;
+using YamlDotNet.RepresentationModel;
+
+namespace Tye.Serialization
+{
+    public static class YamlConfigApplicationHelpers
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:10]: Please no.

File: src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+        public static void HandleConfigApplication(YamlMappingNode? yamlMappingNode, ConfigApplication app)
+        {
+            foreach (var child in yamlMappingNode!.Children)
+            {
+                string? key;
+                switch (child.Key.NodeType)
+                {
+                    case YamlNodeType.Scalar:
+                        key = (child.Key as YamlScalarNode)!.Value;
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:21]: Don't combine an `as` cast with `!`. You want this to be a c-style cast so it will fail with a good error instead of a nullref.

File: src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs (hunk 8 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:10]: What do you mean?

Comment (rynowak) [reply src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:10]: A good rule. You can remove the word `Helpers` and it's suddenly a better name.

Comment (rynowak) [reply src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:10]: `ConfigApplicationParser` would be a good name

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj (hunk 5 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:37]: Directory.Build.targets sets stuff and it's filled in with arcade

File: src/Microsoft.Tye.Core/Serialization/YamlExtensionHelper.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using YamlDotNet.RepresentationModel;
+
+namespace Tye.Serialization
+{
+    public static class YamlExtensionHelper
+    {
+        internal static void HandleExtensionsMapping(YamlSequenceNode yamlSequenceNode, List<Dictionary<string, object>> extensions)
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlExtensionHelper.cs:12]: Can the class be internal instead of the method? It's pretty rare to want a public type with internal methods.

File: src/Microsoft.Tye.Core/Serialization/YamlExtensionHelper.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                            {
+                                case YamlNodeType.Scalar:
+                                    key = (mapping.Key as YamlScalarNode)!.Value;
+                                    break;
+                                default:
+                                    throw new TyeYamlException(mapping.Key.Start,
+                                        CoreStrings.FormatUnexpectedType(YamlNodeType.Scalar.ToString(), mapping.Key.NodeType.ToString()));
+                            }
+
+                            if (key != null)
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlExtensionHelper.cs:33]: why  do we need a null check here? This seems like it should either not happen or be an error if it does.

File: src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs (hunk 2 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs:55]: I like them for having consistent strings and sharing strings between src and test.

What other mechanism would you prefer?

File: src/Microsoft.Tye.Core/Serialization/YamlIngressHelpers.cs [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+                    default:
+                        throw new TyeYamlException(child.Start,
+                            CoreStrings.FormatUnexpectedType(YamlNodeType.Mapping.ToString(), child.NodeType.ToString()));
+                }
+            }
+        }
+
+        private static void HandleIngressMapping(YamlMappingNode? yamlMappingNode, ConfigIngress configIngress)
+        {
+            foreach (var child in yamlMappingNode!.Children)
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlIngressHelpers.cs:33]: nullability tip. If you find you find yourself using defining a nullable parameter and then immediately asserting it's non-null then the parameter should not be nullable. The caller needs to do the checking.

Comment (jkotalik) [reply src/Microsoft.Tye.Core/Serialization/YamlIngressHelpers.cs:33]: I think this is just inconsistent. I have other places where I removed the nullability check. I need to do a second pass of this when I have a fresh mind as there is some low hanging fruit.

File: src/Microsoft.Tye.Core/Serialization/YamlParser.cs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+            {
+                throw new TyeYamlException(node.Start, CoreStrings.FormatExpectedYamlScalar(key));
+            }
+
+            return (node as YamlScalarNode)!.Value;
+        }
+
+        public void Dispose()
+        {
+            _reader.Dispose();
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/Serialization/YamlParser.cs:94]: nitpick: This isn't really right. The problem is that you have two constructors where **we** create the Reader, and thus **own** it, and one constructor where it's passed in, and thus we don't **own** it. 

The best thing to do IMO is not be disposable. If you want to support passing in a string for tests, you don't need dispose for that :)

**Comment by jkotalik** [comment]
> This is great. I left some comments about ideas to clean this up, but they aren't really major, more like whines 👍 I'll leave it up to you which ones you want to deal with.

I like the wines 🍷

File: src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs:55]: I hadn't thought about sharing with tests. Makes good sense 👍

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj (hunk 5 above)
Comment (rynowak) [reply src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:37]: OK. I'm just not sure why we need the attribute and for it to be empty.

File: src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs (hunk 8 above)
Comment (rynowak) [reply src/Microsoft.Tye.Core/Serialization/YamlConfigApplicationHelpers.cs:10]: > What do you mean?

No helpers. Pick better names or don't make classes 😆

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj (hunk 5 above)
Comment (davidfowl) [reply src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:37]: Agree it’s triggering me

Comment (jkotalik) [reply src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:37]: I see now, let me try and remove it.

File: src/Microsoft.Tye.Core/Serialization/YamlParser.cs (hunk 13 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Core/Serialization/YamlParser.cs:94]: The constructor which takes in a TextReader was supposed to be internal/private, thus the parser owns the TextReader.
